ext/openssl/openssl.c

```diff
@@ -5309,48 +5340,704 @@ PHP_FUNCTION(openssl_pkcs7_decrypt)
+	char * p7bfilename = NULL;
+	size_t p7bfilename_len = 0;
+	char * sigfile = NULL;
+	size_t sigfile_len = 0;
+	zend_long encoding = ENCODING_CMS;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "szz|l", &data, &data_len, &crypted, &key, &padding) == FAILURE) {
+	RETVAL_LONG(-1);
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "pl|p!ap!p!p!p!l", &filename, &filename_len,
+				&flags, &signersfilename, &signersfilename_len, &cainfo,
+				&extracerts, &extracerts_len, &datafilename, &datafilename_len,
+				&p7bfilename, &p7bfilename_len,
+				&sigfile, &sigfile_len,&encoding) == FAILURE) {
 		RETURN_THROWS();
 	}
+
+	in = BIO_new_file(filename, PHP_OPENSSL_BIO_MODE_R(flags));
+	if (in == NULL) {
+	  php_openssl_store_errors();
+	  goto clean_exit;
+	}
+	if ( sigfile && CMS_DETACHED ) {
+	  sigbio = BIO_new_file(sigfile, PHP_OPENSSL_BIO_MODE_R(flags));
+
+	  switch (encoding) {
+	  	case ENCODING_PEM:
+			cms = PEM_read_bio_CMS(sigbio,NULL,0,NULL);
+			datain = in;
+			break;
+	     	case ENCODING_DER:
+			cms = d2i_CMS_bio(sigbio,NULL);
+			datain = in;
+			break;
+	    	case ENCODING_CMS:
+	       		cms = SMIME_read_CMS(in, &datain);
+			break;
+	    	default:
+	      		php_error_docref(NULL, E_WARNING, "Unknown encoding");
+			goto clean_exit;
+	  }
+	} else  {
+	    cms = SMIME_read_CMS(in, &datain);
+	}
+
+
+	if (cms == NULL) {
+#if DEBUG_SMIME
+		zend_printf("SMIME_read_CMS failed\n");
+#endif
+		php_openssl_store_errors();
+		goto clean_exit;
+	}
+
+	if (extracerts) {
+		others = php_openssl_load_all_certs_from_file(extracerts);
+		if (others == NULL) {
+			goto clean_exit;
+		}
+	}
+
+	store = php_openssl_setup_verify(cainfo);
+
+	if (!store) {
+		goto clean_exit;
+	}
+	if (php_openssl_open_base_dir_chk(filename)) {
+		goto clean_exit;
+	}
+
+
+	if (datafilename) {
+
+		if (php_openssl_open_base_dir_chk(datafilename)) {
+			goto clean_exit;
+		}
+
+		dataout = BIO_new_file(datafilename, PHP_OPENSSL_BIO_MODE_W(CMS_BINARY));
+		if (dataout == NULL) {
+			php_openssl_store_errors();
+			goto clean_exit;
+		}
+	}
+
+	if (p7bfilename) {
+
+		if (php_openssl_open_base_dir_chk(p7bfilename)) {
+			goto clean_exit;
+		}
+
+		p7bout = BIO_new_file(p7bfilename, PHP_OPENSSL_BIO_MODE_W(CMS_BINARY));
+		if (p7bout == NULL) {
+			php_openssl_store_errors();
+			goto clean_exit;
+		}
+	}
+#if DEBUG_SMIME
+	zend_printf("Calling CMS verify\n");
+#endif
+
+	if (CMS_verify(cms, others, store, datain, dataout, (unsigned int)flags)) {
+
+		RETVAL_TRUE;
+
+		if (signersfilename) {
+			BIO *certout;
+
+			if (php_openssl_open_base_dir_chk(signersfilename)) {
+				goto clean_exit;
+			}
+
+			certout = BIO_new_file(signersfilename, PHP_OPENSSL_BIO_MODE_W(CMS_BINARY));
+			if (certout) {
+				int i;
+				signers = CMS_get0_signers(cms);
+				if (signers != NULL) {
+
+					for (i = 0; i < sk_X509_num(signers); i++) {
+						if (!PEM_write_bio_X509(certout, sk_X509_value(signers, i))) {
+							php_openssl_store_errors();
+							RETVAL_LONG(-1);
```

Comment:
I'm just thinking if `-1` and mixing it with `bool` is good idea. I know that's in PKCS7 but still... Maybe we should just return int always (1 instead of true and 0 instead of false) but not sure if it's worth the inconsistency...

Reply:
That's the issue for me.  I think it's better to return a bool but I am concerned that in this case, it will confuse coders.

Reply:
Yeah it's kind of strange but not sure what's actually better.

Reply:
In thinking about it some more, I think it's bad practice to return a non-boolean, and we shouldn't continue it.  To quote Emmerson, "A foolish consistency is the hobgoblin of small minds."  With this I have made the appropriate changes, run the tests, and will be checking in the code presently.